### PR TITLE
modern cmake config for using gtensor

### PR DESCRIPTION
**Pull request by bd4**

### CODE REVIEW

**Comment by bd4** [comment]
Needs documentation, but I'd like to settle on a final approach first. There is a "better" version on top of this PR here: https://github.com/bd4/gtensor/compare/cmake-config...bd4:pr/cmake-multi-config. It exports `gtensor::gtensor_{host,cuda,hip}`, and defines an alias `gtensor::gtensor`. The supported targets, along with the default alias, is defined at install time. Setting GTENSOR_DEVICE before running find_package allows a different alias to be set. It scratches some corners of cmake I never expected to examine, particularly the hack to make the targets global so they can be aliased at find_package config time. It all works in cmake 3.13 though, so it's not using anything too new.

I think merging this first, then doing a new PR for the modifications, is a reasonable approach. Both are viable and pass tests, there is some value in capturing both approaches in case we decide to switch back at some point.

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,80 +1,71 @@
...
+        LANGUAGES CXX
+        HOMEPAGE_URL https://github.com/wdmapp/gtensor)
 
 option(USE_GTEST_DISCOVER_TESTS "use gtest_discover_tests()" ON)
 set(GTENSOR_DEVICE "cuda" CACHE STRING "Device type 'none', 'cuda', or 'hip'")
 set_property(CACHE GTENSOR_DEVICE PROPERTY STRINGS "none" "cuda" "hip")
+set(GTENSOR_BUILD_EXAMPLES OFF CACHE BOOL "Build example programs")
+
+if (NOT ${CMAKE_PROJECT_NAME} STREQUAL gtensor)
+  set(BUILD_TESTING OFF)
```
[the remark this answers is not in the document]
Comment (bd4) [reply CMakeLists.txt:13]: yes probably better to check this directly instead of messing with `BUILD_TESTING`, I'll change it.

**Comment by bd4** [comment]
I think the common use case is to support `host` plus one device (`hip` or `cuda` or `sycl`). Just like in SYCL, being able to use `host` is useful for debugging, so it makes sense to always have it available. Very few machines other than our own test machines will have multiple types of GPUs - so the use case for that is mostly automated testing...

Another potential use case for multiple backends, where the user might want an easy way to switch between them, is `sycl` + (`hip` or `cuda`). Because there are SYCL implementations that support hip and cuda. This might be an interesting benchmarking exercise. I guess useful more for testing SYCL implementations than for gtensor users.

**Comment by bd4** [comment]
I think the `host` + `cuda` case justifies being able to override the default - install with `cuda` as default, switch app to `host` for testing without having to write your own cmake magic to switch between `gtensor::gtensor` and `gtensor::gtensor_host`.

**Comment by bd4** [comment]
I can also implement the Thrust style, of having a function/macro to import and create the target of your choice, rather than a standard cache variable `-DGTENSOR_DEVICE`. I like the cache variable though, because I imagine it being set via cmake command line, not inside gtensor source, and while apps can implement it themselves, having a standard across projects is nice.

**Comment by germasch** [comment]
> I think the common use case is to support `host` plus one device (`hip` or `cuda` or `sycl`). Just like in SYCL, being able to use `host` is useful for debugging, so it makes sense to always have it available. Very few machines other than our own test machines will have multiple types of GPUs - so the use case for that is mostly automated testing...

Right, I didn't mean to imply that a given install would be limited to only a single supported device, just to a given fixed subset (like host + cuda, or host + cuda + openmp, or whatever is actually compatible).

**Comment by germasch** [comment]
> I think the `host` + `cuda` case justifies being able to override the default - install with `cuda` as default, switch app to `host` for testing without having to write your own cmake magic to switch between `gtensor::gtensor` and `gtensor::gtensor_host`.

Good point. I guess I was thinking one might do this by just setting a macro, ie, still link against gtensor::gtensor (which might mean host+cuda) but add `-DGTENSOR_DEFAULT_DEVICE=host`. Even in that case, I guess having a macro to do that wouldn't hurt.

I. know this is different from the more involved having multiple targets, and then doing an alias for the default. Which is an expression of me not really having figured out what's needed and how best to achieve it.

I think it's definitely good to be able to choose what one wants, but the primary existing option should be to use using `my_gtensor = gt::gtensor<T, space::cuda>` or something like that, with the default for `space` being `device`, and what `device` is aliased to depending on the config that gtensor was built with, and overridable via macro. I think the question then is how much of that needs to be repeated at the cmake level.

One could follow xtensor somewhat in that they have, e.g., a `xtensor_optimize` target which changes compiler flags, so we could have a `gtensor_default_device_host` (or just `gtensor_debug{,host}`? that'd add the `-DGTENSOR_DEFAULT_DEVICE=host`. One advantage to that is that's relatively self-contained, and per target, so e.g. if one is building unit tests, one could build one set for the host and a 2nd set for the device.

**Comment by germasch** [comment]
> I can also implement the Thrust style, of having a function/macro to import and create the target of your choice, rather than a standard cache variable `-DGTENSOR_DEVICE`. I like the cache variable though, because I imagine it being set via cmake command line, not inside gtensor source, and while apps can implement it themselves, having a standard across projects is nice.

So how about this one: If `GTENSOR_DEVICE` is set, just add `-DGTENSOR_DEFAULT_DEVICE=${GTENSOR_DEVICE}` to the gtensor target? (After checking it's actually a supported device, maybe). That seems pretty simple?

**Pull request merged**